### PR TITLE
Move package and container image builds to Nix

### LINKED ISSUE #62: Move package and container image builds to Nix
- Install the Rust toolchain using fenix.
- Use naersk to provide flake outputs that build `fmt`, `clippy` and `test` packages that check the entire workspace.
- Use naersk to provide flake outputs that, for each tool, provide packages to build the package itself and container image.
- Update CI configuration (for now, publishing container images can be ignored).

See https://github.com/DanNixon/satori/pull/26 for an example of almost identical work.

References that might be useful:
- https://github.com/nix-community/fenix
- https://github.com/nix-community/naersk
- https://nixos.wiki/wiki/Rust

### PR DESCRIPTION
**Pull request by Henry-Denny**
Closes #62

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -1,34 +1,81 @@
...
+          buildInputs = buildInputs;
+          packages = with pkgs; [ nix skopeo ];
+        };
+
+        packages = {
+          events-to-histogram = import ./events-to-histogram { inherit pkgs naersk' version git_revision nativeBuildInputs buildInputs; };
+          kafka-daq-report = import ./kafka-daq-report { inherit pkgs naersk' version git_revision nativeBuildInputs buildInputs; };
+          simulator = import ./simulator { inherit pkgs naersk' version git_revision nativeBuildInputs buildInputs; };
+          stream-to-file = import ./stream-to-file { inherit pkgs naersk' version git_revision nativeBuildInputs buildInputs; };
+          streaming-types = import ./streaming-types { inherit pkgs naersk' version git_revision nativeBuildInputs buildInputs; };
```
Comment (DanNixon) [inline flake.nix:49]: `straming-types` is a library used by other crates, so should not appear here.

File: kafka-daq-report/default.nix [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    config = {
+      Entrypoint = [ "${pkgs.tini}/bin/tini" "--" "${entrypoint}/bin/entrypoint" ];
+      ExposedPorts = {
+        "8000/tcp" = {};
+        "9090/tcp" = {};
+      };
+      Env = [
+        "SSL_CERT_FILE=${pkgs.cacert}/etc/ssl/certs/ca-bundle.crt"
+        "HTTP_SERVER_ADDRESS=0.0.0.0:8000"
```
Comment (DanNixon) [inline kafka-daq-report/default.nix:57]:
```suggestion
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -5,11 +5,43 @@ on:
...
+      - uses: actions/checkout@v3
+      - uses: cachix/install-nix-action@v20
+
+      - name: Format
+        run: nix build -L .#fmt
+
+      - name: Clippy
+        run: nix build -L .#clippy
+
+      - name: Start MinIO
```
Comment (DanNixon) [inline .github/workflows/ci.yml:27]: This step is not required.

File: kafka-daq-report/default.nix [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    postInstall = ''
+      wrapProgram $out/bin/kafka-daq-report --prefix PATH : ${pkgs.lib.makeBinPath [ pkgs.ffmpeg ]}
+    '';
+    overrideMain = p: {
+      GIT_REVISION = git_revision;
+    };
+  };
+
+  container-image = let
+    entrypoint = pkgs.writeShellApplication {
```
[the remark this answers is not in the document]
Comment (DanNixon) [reply kafka-daq-report/default.nix:25]: I don't believe there is.

File: events-to-histogram/default.nix [hunk 5]
```diff
@@ -20,37 +20,4 @@
       GIT_REVISION = git_revision;
     };
   };
-
-  container-image = let
```
Comment (DanNixon) [inline events-to-histogram/default.nix]: We still want the container images, they just do not need the entrypoint script.

See https://github.com/DanNixon/satori/blob/main/archiver/default.nix#L24 for an example.

File: kafka-daq-report/default.nix [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  hdf5-joined,
+} : rec {
+  package = naersk'.buildPackage {
+    name = "kafka-daq-report";
+    version = version;
+
+    src = ./..;
+    cargoBuildOptions = x: x ++ ["--package" "kafka-daq-report"];
+
+    nativeBuildInputs = nativeBuildInputs ++ [ pkgs.makeWrapper ];
```
Comment (DanNixon) [inline kafka-daq-report/default.nix:17]:
```suggestion
    nativeBuildInputs = nativeBuildInputs;
```

File: kafka-daq-report/default.nix [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    buildInputs = buildInputs;
+
+    overrideMain = p: {
+      GIT_REVISION = git_revision;
+    };
+
+    HDF5_DIR = "${hdf5-joined}";
```
Comment (DanNixon) [inline kafka-daq-report/default.nix:24]: This is not needed unless the binary depends on `hdf5`.

File: simulator/default.nix [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  buildInputs,
+} : rec {
+  package = naersk'.buildPackage {
+    name = "simulator";
+    version = version;
+
+    src = ./..;
+    cargoBuildOptions = x: x ++ ["--package" "simulator"];
+
+    nativeBuildInputs = nativeBuildInputs ++ [ pkgs.makeWrapper ];
```
Comment (DanNixon) [inline simulator/default.nix:16]:
```suggestion
    nativeBuildInputs = nativeBuildInputs;
```

File: stream-to-file/default.nix [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  hdf5-joined,
+} : rec {
+  package = naersk'.buildPackage {
+    name = "stream-to-file";
+    version = version;
+
+    src = ./..;
+    cargoBuildOptions = x: x ++ ["--package" "stream-to-file"];
+
+    nativeBuildInputs = nativeBuildInputs ++ [ pkgs.makeWrapper ];
```
Comment (DanNixon) [inline stream-to-file/default.nix:17]:
```suggestion
    nativeBuildInputs = nativeBuildInputs;
```

File: trace-archiver/default.nix [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  hdf5-joined,
+} : rec {
+  package = naersk'.buildPackage {
+    name = "trace-archiver";
+    version = version;
+
+    src = ./..;
+    cargoBuildOptions = x: x ++ ["--package" "trace-archiver"];
+
+    nativeBuildInputs = nativeBuildInputs ++ [ pkgs.makeWrapper ];
```
Comment (DanNixon) [inline trace-archiver/default.nix:17]:
```suggestion
    nativeBuildInputs = nativeBuildInputs;
```

File: trace-to-events/default.nix [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  buildInputs,
+} : rec {
+  package = naersk'.buildPackage {
+    name = "trace-to-events";
+    version = version;
+
+    src = ./..;
+    cargoBuildOptions = x: x ++ ["--package" "trace-to-events"];
+
+    nativeBuildInputs = nativeBuildInputs ++ [ pkgs.makeWrapper ];
```
Comment (DanNixon) [inline trace-to-events/default.nix:16]:
```suggestion
    nativeBuildInputs = nativeBuildInputs;
```

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -5,28 +5,61 @@ on:
...
+      - uses: actions/checkout@v3
+      - uses: cachix/install-nix-action@v20
 
       - name: Build
-        run: nix develop --command cargo build --all
+        run: nix build -L .#${{ matrix.package }}.package
 
-      - name: Test
-        run: nix develop --command cargo test --all
+      - name: Build and push container image
```
Comment (DanNixon) [inline .github/workflows/ci.yml:53]: Comment out this step for now, we can add the container builds later on.

**Review by DanNixon — CHANGES_REQUESTED** [review]
Looks good.

One other thing that is worth doing now, could you restructure the `default.nix` files for each component as per https://github.com/DanNixon/satori/commit/49a5ff5d3b6a6be91755bfe7e7a2c8b615cb5640, this will get around the issue we saw earlier with `nix flake show` giving an error.

**Comment by Henry-Denny** [comment]
> Looks good.
> 
> One other thing that is worth doing now, could you restructure the `default.nix` files for each component as per [DanNixon/satori@49a5ff5](https://github.com/DanNixon/satori/commit/49a5ff5d3b6a6be91755bfe7e7a2c8b615cb5640), this will get around the issue we saw earlier with `nix flake show` giving an error.

Thank you. In terms of restructuring, I can't see a difference between the `default.nix` files in https://github.com/DanNixon/satori/commit/49a5ff5d3b6a6be91755bfe7e7a2c8b615cb5640 and the ones currently in https://github.com/STFC-ICD-Research-and-Design/supermusr-data-pipeline/pull/63/commits/f69ff9c4398c30e8eb30939358a4f1e053852bce. The only difference I can spot is the `entryPoint` and `postInstall` code, but I thought that these were left out intentionally.

**Comment by DanNixon** [comment]
> > Looks good.
> > One other thing that is worth doing now, could you restructure the `default.nix` files for each component as per [DanNixon/satori@49a5ff5](https://github.com/DanNixon/satori/commit/49a5ff5d3b6a6be91755bfe7e7a2c8b615cb5640), this will get around the issue we saw earlier with `nix flake show` giving an error.
> 
> Thank you. In terms of restructuring, I can't see a difference between the `default.nix` files in [DanNixon/satori@49a5ff5](https://github.com/DanNixon/satori/commit/49a5ff5d3b6a6be91755bfe7e7a2c8b615cb5640) and the ones currently in [f69ff9c](https://github.com/STFC-ICD-Research-and-Design/supermusr-data-pipeline/commit/f69ff9c4398c30e8eb30939358a4f1e053852bce). The only difference I can spot is the `entryPoint` and `postInstall` code, but I thought that these were left out intentionally.

Looking at the commit I linked to in isolation should show the changes I am referring to, specifically, it makes everything under `packages` a derivation, i.e. what is currently `trace-archiver.package` and `trace-archiver.container-image become `trace-archiver` and `trace-arhciver-container-image`.

**Review by DanNixon — APPROVED** [review]
Looks good, there will be some follow up work to do, but as it stands this is a good amount of progress and all the foundations and essentials are in place.

**Pull request merged**